File: ForgeHub/backend/src/main/java/id/jakforge/forgehub/api/AssetsApi.java
package id.jakforge.forgehub.api;

import id.jakforge.forgehub.data.Db;
import id.jakforge.forgehub.security.CurrentUser;
import id.jakforge.forgehub.security.SecretBox;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.DeleteMapping;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RestController;

import java.util.HashMap;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.Set;
import java.util.UUID;

/**
 * Aset dan kredensial.
 *
 * <p>Keduanya menyimpan nilai yang harus bisa DIBACA KEMBALI oleh robot, bukan
 * sekadar diringkas: robot perlu kata sandi aslinya untuk masuk ke aplikasi
 * yang diotomasi. Karena itu yang dipakai adalah penyandian dua arah
 * ({@link SecretBox}), bukan hash.
 *
 * <p>Nilai rahasia dibuka HANYA di endpoint {@code /value}. Daftar biasa
 * mengembalikan penanda "ada isinya", bukan isinya — supaya membuka halaman
 * Assets tidak menaburkan seluruh kata sandi ke dalam log peramban, riwayat
 * proksi, dan tangkapan layar.
 */
@RestController
@RequestMapping("/api")
public class AssetsApi {

    private static final Set<String> TIPE_SAH = Set.of("Text", "Integer", "Bool", "Credential", "Secret");
    private static final Set<String> TIPE_RAHASIA = Set.of("Credential", "Secret");

    private final Db db;
    private final SecretBox secrets;

    public AssetsApi(Db db, SecretBox secrets) {
        this.db = db;
        this.secrets = secrets;
    }

    // -----------------------------------------------------------------
    // Aset
    // -----------------------------------------------------------------

    @GetMapping("/assets")
    public List<Map<String, Object>> aset() {
        return db.rows("""
                SELECT id, name, type, scope, description, created_at, updated_at,
                       CASE WHEN type IN ('Credential', 'Secret') THEN NULL ELSE value_text END AS value_text,
                       CASE WHEN value_text IS NULL OR value_text = '' THEN FALSE ELSE TRUE END AS has_value
                  FROM assets
                 WHERE tenant_id = ?
                 ORDER BY name
                """, CurrentUser.get().tenantId());
    }

    @PostMapping("/assets")
    public ResponseEntity<?> simpanAset(@RequestBody(required = false) Map<String, Object> body) {
        String name = Badan.nama(body, "name");

        if (name == null) {
            return ResponseEntity.badRequest().body(Map.of("error", "Nama aset wajib diisi."));
        }

        String tipe = Badan.teks(body, "type", "Text");
        String nilai = Badan.teks(body, "value");

        if (!TIPE_SAH.contains(tipe)) {
            return ResponseEntity.badRequest().body(Map.of(
                    "error", "Tipe aset tidak dikenal: '" + tipe + "'."));
        }

        // Nilai bertipe angka dan boolean diperiksa DI SINI, bukan dibiarkan
        // meledak nanti di dalam robot yang sedang berjalan. Kegagalan di sini
        // dilihat orang yang baru saja mengetiknya; kegagalan di sana dilihat
        // sebagai automasi yang berhenti di tengah malam.
        if ("Integer".equals(tipe) && nilai != null && !nilai.isEmpty()) {
            try {
                Long.parseLong(nilai.trim());
            } catch (NumberFormatException e) {
                return ResponseEntity.badRequest().body(Map.of(
                        "error", "Aset bertipe Integer harus berisi bilangan bulat."));
            }
        }

        if ("Bool".equals(tipe) && nilai != null && !nilai.isEmpty()) {
            String v = nilai.trim();
            if (!v.equalsIgnoreCase("true") && !v.equalsIgnoreCase("false")) {
                return ResponseEntity.badRequest().body(Map.of(
                        "error", "Aset bertipe Bool harus berisi true atau false."));
            }
        }

        String tersimpan = TIPE_RAHASIA.contains(tipe) ? secrets.protect(nilai) : nilai;

        UUID tenantId = CurrentUser.get().tenantId();

        Object adaId = db.scalar("SELECT id FROM assets WHERE tenant_id = ? AND name = ?", tenantId, name);

        if (adaId != null) {
            db.exec("""
                    UPDATE assets
                       SET type = ?, value_text = ?, description = ?, scope = ?, updated_at = now()
                     WHERE tenant_id = ? AND name = ?
                    """, tipe, tersimpan, Badan.teks(body, "description"),
                    Badan.teks(body, "scope", "Global"), tenantId, name);
        } else {
            db.exec("""
                    INSERT INTO assets (id, tenant_id, name, type, value_text, description, scope,
                                        created_at, updated_at)
                    VALUES (?, ?, ?, ?, ?, ?, ?, now(), now())
                    """, Db.newId(), tenantId, name, tipe, tersimpan,
                    Badan.teks(body, "description"), Badan.teks(body, "scope", "Global"));
        }

        Map<String, Object> hasil = new LinkedHashMap<>();
        hasil.put("ok", true);
        hasil.put("created", adaId == null);

        return ResponseEntity.ok(hasil);
    }

    /** Pengambilan nilai untuk robot. Aset rahasia dibuka di sini SAJA. */
    @GetMapping("/assets/{name}/value")
    public ResponseEntity<?> nilaiAset(@PathVariable String name) {
        Map<String, Object> row = db.row(
                "SELECT type, value_text FROM assets WHERE tenant_id = ? AND name = ?",
                CurrentUser.get().tenantId(), name);

        if (row == null) {
            return ResponseEntity.status(HttpStatus.NOT_FOUND)
                    .body(Map.of("error", "Aset '" + name + "' tidak ada."));
        }

        String tipe = (String) row.get("type");
        String mentah = (String) row.get("valueText");
        String nilai = TIPE_RAHASIA.contains(tipe) ? secrets.unprotect(mentah) : mentah;

        // HashMap, bukan Map.of: nilainya boleh null, dan Map.of melempar
        // NullPointerException untuk nilai null. Aset yang ada tapi kosong
        // adalah keadaan yang sah.
        Map<String, Object> hasil = new HashMap<>();
        hasil.put("name", name);
        hasil.put("type", tipe);
        hasil.put("value", nilai);

        return ResponseEntity.ok(hasil);
    }

    @DeleteMapping("/assets/{name}")
    public ResponseEntity<?> hapusAset(@PathVariable String name) {
        int terhapus = db.exec("DELETE FROM assets WHERE tenant_id = ? AND name = ?",
                CurrentUser.get().tenantId(), name);

        return terhapus == 0
                ? ResponseEntity.status(HttpStatus.NOT_FOUND).body(Map.of("error", "Aset tidak ada."))
                : ResponseEntity.ok(Map.of("ok", true));
    }

    // -----------------------------------------------------------------
    // Kredensial
    // -----------------------------------------------------------------

    /** Daftar kredensial TANPA kata sandinya — kolomnya tidak ikut dipilih. */
    @GetMapping("/credentials")
    public List<Map<String, Object>> kredensial() {
        return db.rows("""
                SELECT id, name, username, description, created_at
                  FROM credentials
                 WHERE tenant_id = ?
                 ORDER BY name
                """, CurrentUser.get().tenantId());
    }

    @PostMapping("/credentials")
    public ResponseEntity<?> simpanKredensial(@RequestBody(required = false) Map<String, Object> body) {
        String name = Badan.nama(body, "name");

        if (name == null) {
            return ResponseEntity.badRequest().body(Map.of("error", "Nama kredensial wajib diisi."));
        }

        UUID tenantId = CurrentUser.get().tenantId();
        String tersandi = secrets.protect(Badan.teks(body, "password"));

        if (db.exists("SELECT count(*) FROM credentials WHERE tenant_id = ? AND name = ?", tenantId, name)) {
            // COALESCE pada password_enc: menyunting keterangan tanpa mengisi
            // ulang kata sandinya tidak boleh MENGHAPUS kata sandi yang ada.
            // Ini beda dari sisi .NET, yang menimpanya dengan null — dan
            // akibatnya baru terlihat saat robot berikutnya gagal masuk.
            db.exec("""
                    UPDATE credentials
                       SET username = ?, password_enc = COALESCE(?, password_enc), description = ?
                     WHERE tenant_id = ? AND name = ?
                    """, Badan.teks(body, "username"), tersandi,
                    Badan.teks(body, "description"), tenantId, name);
        } else {
            db.exec("""
                    INSERT INTO credentials (id, tenant_id, name, username, password_enc, description, created_at)
                    VALUES (?, ?, ?, ?, ?, ?, now())
                    """, Db.newId(), tenantId, name, Badan.teks(body, "username"),
                    tersandi, Badan.teks(body, "description"));
        }

        return ResponseEntity.ok(Map.of("ok", true));
    }

    @GetMapping("/credentials/{name}/value")
    public ResponseEntity<?> nilaiKredensial(@PathVariable String name) {
        Map<String, Object> row = db.row(
                "SELECT username, password_enc FROM credentials WHERE tenant_id = ? AND name = ?",
                CurrentUser.get().tenantId(), name);

        if (row == null) {
            return ResponseEntity.status(HttpStatus.NOT_FOUND)
                    .body(Map.of("error", "Kredensial tidak ada."));
        }

        Map<String, Object> hasil = new HashMap<>();
        hasil.put("username", row.get("username"));
        hasil.put("password", secrets.unprotect((String) row.get("passwordEnc")));

        return ResponseEntity.ok(hasil);
    }

    @DeleteMapping("/credentials/{name}")
    public ResponseEntity<?> hapusKredensial(@PathVariable String name) {
        int terhapus = db.exec("DELETE FROM credentials WHERE tenant_id = ? AND name = ?",
                CurrentUser.get().tenantId(), name);

        return terhapus == 0
                ? ResponseEntity.status(HttpStatus.NOT_FOUND).body(Map.of("error", "Kredensial tidak ada."))
                : ResponseEntity.ok(Map.of("ok", true));
    }
}

File: ForgeHub/server/Api/AssetsApi.cs
using ForgeHub.Auth;
using ForgeHub.Data;

namespace ForgeHub.Api;

public static class AssetsApi
{
    /// <summary>Batas ukuran satu berkas di gudang: 32 MB.</summary>
    private const long MaxFileBytes = 32L * 1024 * 1024;

    public static void MapAssets(this WebApplication app)
    {
        // --------------------------------------------------------------
        // Aset bertipe Credential/Secret TIDAK menampilkan nilainya di daftar;
        // yang muncul hanya penanda bahwa isinya ada.
        app.MapGet("/api/assets", (HttpContext context) =>
        {
            using var connection = context.Db(out var tenantId);

            var rows = Sql.Rows(connection,
                @"SELECT id, name, type, scope, description, created_at, updated_at,
                         CASE WHEN type IN ('Credential', 'Secret') THEN NULL ELSE value_text END AS value_text,
                         CASE WHEN value_text IS NULL OR value_text = '' THEN 0 ELSE 1 END AS has_value
                    FROM assets
                   WHERE tenant_id = @p0
                   ORDER BY name",
                tenantId);

            return Results.Ok(rows);
        });

        // --------------------------------------------------------------
        app.MapPost("/api/assets", async (HttpContext context) =>
        {
            var body = await context.Body();
            var name = body.Str("name");

            if (string.IsNullOrWhiteSpace(name))
                return Results.BadRequest(new { error = "Nama aset wajib diisi." });

            var type = body.Str("type", "Text");
            var value = body.Str("value");

            var allowed = new[] { "Text", "Integer", "Bool", "Credential", "Secret" };
            if (!allowed.Contains(type))
                return Results.BadRequest(new { error = "Tipe aset tidak dikenal: '" + type + "'." });

            // Nilai bertipe angka dan boolean diperiksa di sini, bukan dibiarkan
            // meledak nanti di dalam robot yang sedang berjalan.
            if (type == "Integer" && !string.IsNullOrEmpty(value) && !long.TryParse(value, out _))
                return Results.BadRequest(new { error = "Aset bertipe Integer harus berisi bilangan bulat." });

            if (type == "Bool" && !string.IsNullOrEmpty(value) &&
                !bool.TryParse(value, out _))
                return Results.BadRequest(new { error = "Aset bertipe Bool harus berisi true atau false." });

            var stored = type is "Credential" or "Secret" ? Secrets.Protect(value) : value;

            using var connection = context.Db(out var tenantId);
            var now = Sql.Now();

            var existing = (string)Sql.Scalar(connection,
                "SELECT id FROM assets WHERE tenant_id = @p0 AND name = @p1", tenantId, name);

            if (existing != null)
            {
                Sql.Exec(connection,
                    @"UPDATE assets
                         SET type = @p0, value_text = @p1, description = @p2, scope = @p3, updated_at = @p4
                       WHERE id = @p5",
                    type, stored, body.Str("description"), body.Str("scope", "Global"), now, existing);
            }
            else
            {
                Sql.Exec(connection,
                    @"INSERT INTO assets (id, tenant_id, name, type, value_text, description, scope, created_at, updated_at)
                      VALUES (@p0, @p1, @p2, @p3, @p4, @p5, @p6, @p7, @p7)",
                    Sql.NewId(), tenantId, name, type, stored,
                    body.Str("description"), body.Str("scope", "Global"), now);
            }

            return Results.Ok(new { ok = true, created = existing == null });
        });

        // --------------------------------------------------------------
        // Pengambilan nilai untuk robot. Aset rahasia dibuka di sini SAJA.
        app.MapGet("/api/assets/{name}/value", (HttpContext context, string name) =>
        {
            using var connection = context.Db(out var tenantId);

            var row = Sql.Row(connection,
                "SELECT type, value_text FROM assets WHERE tenant_id = @p0 AND name = @p1", tenantId, name);

            if (row == null) return Results.NotFound(new { error = "Aset '" + name + "' tidak ada." });

            var type = (string)row["type"];
            var raw = (string)row["valueText"];
            var value = type is "Credential" or "Secret" ? Secrets.Unprotect(raw) : raw;

            return Results.Ok(new { name, type, value });
        });

        app.MapDelete("/api/assets/{name}", (HttpContext context, string name) =>
        {
            using var connection = context.Db(out var tenantId);

            var removed = Sql.Exec(connection,
                "DELETE FROM assets WHERE tenant_id = @p0 AND name = @p1", tenantId, name);

            return removed == 0 ? Results.NotFound(new { error = "Aset tidak ada." }) : Results.Ok(new { ok = true });
        });

        MapBuckets(app);
    }

    // ------------------------------------------------------------------

    private static void MapBuckets(WebApplication app)
    {
        app.MapGet("/api/buckets", (HttpContext context) =>
        {
            using var connection = context.Db(out var tenantId);

            var rows = Sql.Rows(connection,
                @"SELECT b.id, b.name, b.description, b.created_at,
                         (SELECT COUNT(*) FROM bucket_files f
                           WHERE f.tenant_id = b.tenant_id AND f.bucket_name = b.name) AS file_count,
                         (SELECT COALESCE(SUM(f.size_bytes), 0) FROM bucket_files f
                           WHERE f.tenant_id = b.tenant_id AND f.bucket_name = b.name) AS total_bytes
                    FROM buckets b
                   WHERE b.tenant_id = @p0
                   ORDER BY b.name",
                tenantId);

            return Results.Ok(rows);
        });

        app.MapPost("/api/buckets", async (HttpContext context) =>
        {
            var body = await context.Body();
            var name = body.Str("name");

            if (string.IsNullOrWhiteSpace(name))
                return Results.BadRequest(new { error = "Nama gudang wajib diisi." });

            using var connection = context.Db(out var tenantId);

            if (Sql.Count(connection,
                    "SELECT COUNT(*) FROM buckets WHERE tenant_id = @p0 AND name = @p1", tenantId, name) > 0)
                return Results.Conflict(new { error = "Gudang '" + name + "' sudah ada." });

            Sql.Exec(connection,
                "INSERT INTO buckets (id, tenant_id, name, description, created_at) VALUES (@p0, @p1, @p2, @p3, @p4)",
                Sql.NewId(), tenantId, name, body.Str("description"), Sql.Now());

            return Results.Ok(new { ok = true });
        });

        app.MapDelete("/api/buckets/{name}", (HttpContext context, string name) =>
        {
            using var connection = context.Db(out var tenantId);

            Sql.Exec(connection, "DELETE FROM bucket_files WHERE tenant_id = @p0 AND bucket_name = @p1", tenantId, name);

            var removed = Sql.Exec(connection,
                "DELETE FROM buckets WHERE tenant_id = @p0 AND name = @p1", tenantId, name);

            return removed == 0 ? Results.NotFound(new { error = "Gudang tidak ada." }) : Results.Ok(new { ok = true });
        });

        // --------------------------------------------------------------
        app.MapGet("/api/buckets/{name}/files", (HttpContext context, string name) =>
        {
            using var connection = context.Db(out var tenantId);

            var rows = Sql.Rows(connection,
                @"SELECT id, file_name, content_type, size_bytes, uploaded_by, uploaded_at
                    FROM bucket_files
                   WHERE tenant_id = @p0 AND bucket_name = @p1
                   ORDER BY uploaded_at DESC",
                tenantId, name);

            return Results.Ok(rows);
        });

        app.MapPost("/api/buckets/{name}/files", async (HttpContext context, string name) =>
        {
            var body = await context.Body();

            var fileName = body.Str("fileName");
            if (string.IsNullOrWhiteSpace(fileName))
                return Results.BadRequest(new { error = "fileName wajib diisi." });

            // Nama berkas datang dari luar dan dipakai sebagai nama unduhan.
            // Pemisah jalur dibuang supaya tidak ada yang bisa menulis "../"
            // ke dalamnya dan mengubah tempat berkas itu mendarat di komputer
            // orang yang mengunduhnya.
            fileName = Path.GetFileName(fileName.Replace('\\', '/'));

            if (string.IsNullOrWhiteSpace(fileName))
                return Results.BadRequest(new { error = "fileName tidak sah." });

            byte[] content;
            try
            {
                content = Convert.FromBase64String(body.Str("contentBase64") ?? "");
            }
            catch (FormatException)
            {
                return Results.BadRequest(new { error = "contentBase64 bukan base64 yang sah." });
            }

            if (content.LongLength > MaxFileBytes)
                return Results.BadRequest(new
                {
                    error = "Berkas terlalu besar. Batasnya " + (MaxFileBytes / 1024 / 1024) + " MB.",
                });

            using var connection = context.Db(out var tenantId);

            if (Sql.Count(connection,
                    "SELECT COUNT(*) FROM buckets WHERE tenant_id = @p0 AND name = @p1", tenantId, name) == 0)
                return Results.NotFound(new { error = "Gudang '" + name + "' tidak ada." });

            Sql.Exec(connection,
                "DELETE FROM bucket_files WHERE tenant_id = @p0 AND bucket_name = @p1 AND file_name = @p2",
                tenantId, name, fileName);

            Sql.Exec(connection,
                @"INSERT INTO bucket_files
                    (id, tenant_id, bucket_name, file_name, content_type, size_bytes, uploaded_by, uploaded_at, content)
                  VALUES (@p0, @p1, @p2, @p3, @p4, @p5, @p6, @p7, @p8)",
                Sql.NewId(), tenantId, name, fileName,
                body.Str("contentType", "application/octet-stream"),
                content.LongLength, context.User()?.Username, Sql.Now(), content);

            return Results.Ok(new { ok = true, fileName, sizeBytes = content.LongLength });
        });

        app.MapGet("/api/buckets/{name}/files/{id}/content", (HttpContext context, string name, string id) =>
        {
            using var connection = context.Db(out var tenantId);

            var row = Sql.Row(connection,
                @"SELECT file_name, content_type, content FROM bucket_files
                   WHERE tenant_id = @p0 AND bucket_name = @p1 AND id = @p2",
                tenantId, name, id);

            if (row == null || row["content"] == null)
                return Results.NotFound(new { error = "Berkas tidak ada." });

            return Results.File((byte[])row["content"],
                (string)row["contentType"] ?? "application/octet-stream",
                (string)row["fileName"]);
        });

        app.MapDelete("/api/buckets/{name}/files/{id}", (HttpContext context, string name, string id) =>
        {
            using var connection = context.Db(out var tenantId);

            var removed = Sql.Exec(connection,
                "DELETE FROM bucket_files WHERE tenant_id = @p0 AND bucket_name = @p1 AND id = @p2",
                tenantId, name, id);

            return removed == 0 ? Results.NotFound(new { error = "Berkas tidak ada." }) : Results.Ok(new { ok = true });
        });
    }
}
